filesystem/android/src/main/java/com/capacitorjs/plugins/filesystem/Filesystem.java

```diff
@@ -303,81 +307,123 @@ public void copyRecursively(File src, File dst) throws IOException {
         }
     }
 
-    public JSObject downloadFile(PluginCall call, Bridge bridge, HttpRequestHandler.ProgressEmitter emitter)
-        throws IOException, URISyntaxException, JSONException {
+    public void downloadFile(PluginCall call, Bridge bridge, HttpRequestHandler.ProgressEmitter emitter,
+            FilesystemDownloadCallback callback) throws IOException, URISyntaxException, JSONException {
         String urlString = call.getString("url", "");
-        JSObject headers = call.getObject("headers", new JSObject());
-        JSObject params = call.getObject("params", new JSObject());
-        Integer connectTimeout = call.getInt("connectTimeout");
-        Integer readTimeout = call.getInt("readTimeout");
-        Boolean disableRedirects = call.getBoolean("disableRedirects");
-        Boolean shouldEncode = call.getBoolean("shouldEncodeUrlParams", true);
-        Boolean progress = call.getBoolean("progress", false);
-
-        String method = call.getString("method", "GET").toUpperCase(Locale.ROOT);
-        String path = call.getString("path");
-        String directory = call.getString("directory", Environment.DIRECTORY_DOWNLOADS);
-
-        final URL url = new URL(urlString);
-        final File file = getFileObject(path, directory);
-
-        HttpRequestHandler.HttpURLConnectionBuilder connectionBuilder = new HttpRequestHandler.HttpURLConnectionBuilder()
-            .setUrl(url)
-            .setMethod(method)
-            .setHeaders(headers)
-            .setUrlParams(params, shouldEncode)
-            .setConnectTimeout(connectTimeout)
-            .setReadTimeout(readTimeout)
-            .setDisableRedirects(disableRedirects)
-            .openConnection();
-
-        CapacitorHttpUrlConnection connection = connectionBuilder.build();
-
-        connection.setSSLSocketFactory(bridge);
-
-        InputStream connectionInputStream = connection.getInputStream();
-        FileOutputStream fileOutputStream = new FileOutputStream(file, false);
-
-        String contentLength = connection.getHeaderField("content-length");
-        int bytes = 0;
-        int maxBytes = 0;
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        Handler handler = new Handler(Looper.getMainLooper());
+
+        executor.execute(() -> {
+            AsyncTaskResult result = doDownloadInBackground(urlString, call, bridge, emitter);
+            handler.post(() -> {
+                if (result.error != null) {
+                    callback.onError(result.error);
+                } else {
+                    callback.onSuccess(result.result);
+                }
+            });
```

Comment:
It seems to me that it is unnecessary to have this `AsyncTaskResult`, since it either just wraps the result or error if it exists.

I feel like we could lift the try-catch out of `doDownloadInBackground`, and have it return the `JSObject` directly

```suggestion
            try {
                JSObject result = doDownloadInBackground(urlString, call, bridge, emitter);
                handler.post(() -> callback.onSuccess(result));
            } catch (Exception error) {
                handler.post(() -> callback.onError(error));
            }
```

Also, if you do this change, add the `throws IOException, URISyntaxException, JSONException` to the `doDownloadInBackground` signature.

What do you think?